src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+        }
+
+        public static IEnumerable<TSource> Prepend<TSource>(this IEnumerable<TSource> source, TSource element)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            AppendPrependIterator<TSource> ap = source as AppendPrependIterator<TSource>;
+            if (ap != null)
+            {
+                return ap.Prepend(element);
+            }
+
+            return new AppendPrepend1Iterator<TSource>(source, element, false);
+        }
+
+        private abstract class AppendPrependIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        {
+            protected readonly IEnumerable<TSource> _source;
+            private IEnumerator<TSource> _enumerator;
+
+            protected AppendPrependIterator(IEnumerable<TSource> source)
+            {
+                Debug.Assert(source != null);
+                _source = source;
+            }
+
+            protected void GetSourceEnumerator()
+            {
+                Debug.Assert(_enumerator == null);
+                _enumerator = _source.GetEnumerator();
+            }
+
+            protected bool LoadFromEnumerator()
+            {
+                if (_enumerator.MoveNext())
+                {
+                    _current = _enumerator.Current;
+                    return true;
+                }
+
+                DisposeEnumerator();
+                return false;
+            }
+
+            private void DisposeEnumerator()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+            }
+
+            public sealed override void Dispose()
+            {
+                DisposeEnumerator();
+
+                base.Dispose();
+            }
+
+            public abstract AppendPrependIterator<TSource> Append(TSource element);
+
+            public abstract AppendPrependIterator<TSource> Prepend(TSource element);
+
+            public abstract TSource[] ToArray();
+
+            public abstract List<TSource> ToList();
+
+            public abstract int GetCount(bool onlyIfCheap);
+        }
+
+        private sealed class AppendPrepend1Iterator<TSource> : AppendPrependIterator<TSource>
+        {
+            private readonly TSource _element;
+            private readonly bool _appending;
+
+            public AppendPrepend1Iterator(IEnumerable<TSource> source, TSource element, bool appending)
+                : base(source)
+            {
+                _element = element;
+                _appending = appending;
+            }
+
+            public override Iterator<TSource> Clone() => new AppendPrepend1Iterator<TSource>(_source, _element, _appending);
+
+            public override bool MoveNext()
+            {
+                switch (_state)
+                {
+                    case 1:
```

Comment:
It's a linear progression through numbered states. I think the most readable labels are simply 1, 2, 3 etc.